### PR TITLE
feat: add qrcode

### PR DESCRIPTION
**Pull request by selicens**
First of all, thank you for your contribution! 😄

New feature please send pull request to feature branch, and rest to master branch. Pull request will be merged after one of collaborators approve. Please makes sure that these form are filled before submitting your pull request, thank you!

[[中文版模板 / Chinese template](https://github.com/vueComponent/ant-design-vue/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Branch merge
- [ ] Other (about what?)

### What's the background?

> 1. Describe the source of requirement.
> 2. Resolve what problem.
> 3. Related issue link.

### API Realization (Optional if not new feature)

> 1. Basic thought of solution and other optional proposal.
> 2. List final API realization and usage sample.
> 3. GIF or snapshot should be provided if includes UI/interactive modification.

### What's the effect? (Optional if not new feature)

> 1. Does this PR affect user? Which part will be affected?
> 2. What will say in changelog?
> 3. Does this PR contains potential break change or other risk?

### Changelog description (Optional if not new feature)

> 1. English description
> 2. Chinese description (optional)

### Self Check before Merge

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### Additional Plan? (Optional if not new feature)

> If this PR related with other PR or following info. You can type here.

### CODE REVIEW

**Comment by selicens** [comment]
qrcode组件相关：因为antdesign是使用了qrcode.react的库，所以参照jeecgboot-vue3的二维码组件使用了qrcode这个库来实现，npm地址：[https://www.npmjs.com/package/qrcode](https://www.npmjs.com/package/qrcode)，github仓库：[github.com/soldair/node-qrcode](https://github.com/soldair/node-qrcode)，二维码中心icon使用了canvas 2D API的drawImage()方法在canvas中绘制

File: components/qrcode/index.tsx [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    });
+    return () => (
+      <>
+        <canvas ref={qrcodeCanvasRef} />
+      </>
+    );
+  },
+});
+const QRCode = defineComponent({
+  name: 'AQrCode',
```
Comment (tangjinzhou) [inline components/qrcode/index.tsx:123]: AQrcode  这样吧，使用的时候方便 a-qrcode

File: components/qrcode/index.en-US.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+| :-- | :-- | :-- | :-- |
+| value | scanned link | string | - |
+| icon | include image url (only image link are supported) | string | - |
+| size | QRCode size | number | 128 |
+| iconSize | include image size | number | 32 |
+| color | QRCode Color | string | `#000` |
+| bordered | Whether has border style | boolean | `true` |
+| errorLevel | Error Code Level | `'L'` \| `'M'` \| `'Q'` \| `'H'` | `'M'` |
+| status | QRCode status | `active` \| `expired` \| `loading ` | `active` |
+| onRefresh | callback | `() => void` | - |
```
Comment (tangjinzhou) [inline components/qrcode/index.en-US.md:24]: 事件单独列出

File: components/qrcode/interface.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  size?: number;
+  color?: string;
+  style?: CSSProperties;
+  includeMargin?: boolean;
+  imageSettings?: ImageSettings;
+}
+
+export type QRPropsCanvas = QRProps;
+
+export interface QRCodeProps extends QRProps {
```
Comment (tangjinzhou) [inline components/qrcode/interface.ts:23]: 直接用 组件的 QRCodeProps 就好了，不用额外在写一份了

File: components/qrcode/index.tsx [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+                )}
+                <QRCodeCanvas {...qrcodeCanvasProps} />
+              </div>
+            );
+          }}
+        />,
+      );
+    };
+  },
+  methods: {
```
Comment (tangjinzhou) [inline components/qrcode/index.tsx:186]: 不要混用 composition api  和 options api

File: components/qrcode/index.tsx [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+      return null;
+    }
+    if (process.env.NODE_ENV !== 'production') {
+      warning(
+        !(icon.value && errorLevel.value === 'L'),
+        'QRCode',
+        'ErrorLevel `L` is not recommended to be used with `icon`, for scanning result would be affected by low level.',
+      );
+    }
+    const qrcodeCanvasProps: QRCodeCanvasProps = {
```
Comment (tangjinzhou) [inline components/qrcode/index.tsx:144]: 响应式丢失了

File: components/qrcode/index.tsx [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+    const [wrapSSR, hashId] = useStyle(prefixCls);
+    const { value, errorLevel, size, icon, iconSize, color, status, bordered } = toRefs(props);
+    const pre = prefixCls.value;
+    if (!props.value) {
+      if (process.env.NODE_ENV !== 'production') {
+        warning(false, 'QRCode', 'need to receive `value` props');
+      }
+      return null;
+    }
+    if (process.env.NODE_ENV !== 'production') {
```
Comment (tangjinzhou) [inline components/qrcode/index.tsx:137]: 我们是新组件，没有历史包袱，不推荐的属性就不用加进去了

File: components/qrcode/index.tsx [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+      value: value.value,
+      errorLevel: errorLevel.value,
+      size: size.value - 26,
+      icon: icon.value,
+      iconSize: iconSize.value,
+      color: color.value,
+    };
+    return () => {
+      return wrapSSR(
+        <LocaleReceiver
```
Comment (tangjinzhou) [inline components/qrcode/index.tsx:154]: 参考其他组件用  useLocaleReceiver

File: components/qrcode/index.tsx [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+                class={classNames(hashId.value, pre, {
+                  [`${prefixCls}-borderless`]: !bordered.value,
+                })}
+              >
+                {status.value !== 'active' && (
+                  <div class={classNames(`${pre}-mask`)}>
+                    {status.value === 'loading' && <Spin />}
+                    {status.value === 'expired' && (
+                      <>
+                        <p class={classNames(`${pre}-expired`)}>二维码已过期</p>
```
Comment (tangjinzhou) [inline components/qrcode/index.tsx:169]: ？  怎么还出现了 中文

Comment (tangjinzhou) [reply components/qrcode/index.tsx:169]: 国际化内容应该在  components/locale 下面添加

**Review by tangjinzhou — CHANGES_REQUESTED** [review]

File: components/qrcode/index.tsx (hunk 1 above)
Comment (selicens) [reply components/qrcode/index.tsx:123]: 1

File: components/qrcode/index.tsx [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+const QRCode = defineComponent({
+  name: 'AQrCode',
+  props: initDefaultProps(qrcodeProps(), {}),
+  emits: ['refresh'],
+  setup(props, { emit }) {
+    const { prefixCls } = useConfigInject('qrcode', props);
+    const [wrapSSR, hashId] = useStyle(prefixCls);
+    const { value, errorLevel, size, icon, iconSize, color, status, bordered } = toRefs(props);
+    const pre = prefixCls.value;
+    if (!props.value) {
```
Comment (tangjinzhou) [inline components/qrcode/index.tsx:131]: 属性上面声明 value 为必填就好了

**Review by tangjinzhou — CHANGES_REQUESTED** [review]

File: components/qrcode/index.tsx [hunk 10]
```diff
@@ -120,73 +110,53 @@ const QRCodeCanvas = defineComponent({
...
   emits: ['refresh'],
-  setup(props, { emit }) {
+  setup(props, { emit, expose }) {
+    const [locale] = useLocaleReceiver('QRCode', defaultLocale.QRCode);
     const { prefixCls } = useConfigInject('qrcode', props);
     const [wrapSSR, hashId] = useStyle(prefixCls);
-    const { value, errorLevel, size, icon, iconSize, color, status, bordered } = toRefs(props);
+    const [, token] = useToken();
+    const { value, size, status, bordered } = toRefs(props);
     const pre = prefixCls.value;
```
Comment (tangjinzhou) [inline components/qrcode/index.tsx:122]: 在 setup 里面直接拿到的 .value  值，赋值给一个新的变量，这个新变量都是没有响应式的

File: components/qrcode/index.tsx [hunk 11]
```diff
@@ -120,73 +110,53 @@ const QRCodeCanvas = defineComponent({
...
   props: initDefaultProps(qrcodeProps(), {}),
   emits: ['refresh'],
-  setup(props, { emit }) {
+  setup(props, { emit, expose }) {
+    const [locale] = useLocaleReceiver('QRCode', defaultLocale.QRCode);
     const { prefixCls } = useConfigInject('qrcode', props);
     const [wrapSSR, hashId] = useStyle(prefixCls);
-    const { value, errorLevel, size, icon, iconSize, color, status, bordered } = toRefs(props);
+    const [, token] = useToken();
+    const { value, size, status, bordered } = toRefs(props);
```
Comment (tangjinzhou) [inline components/qrcode/index.tsx:121]: 这几个值没必要转一次吧，都没做什么操作啊 
直接放到 return 里面解构就好了

**Review by tangjinzhou — CHANGES_REQUESTED** [review]

File: components/qrcode/index.tsx (hunk 11 above)
Comment (selicens) [reply components/qrcode/index.tsx:121]: 1

**Pull request merged**